util/conversion/conversion.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion
+
+import (
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/json"
+)
+
+const (
+	DataAnnotation = "cluster.x-k8s.io/conversion-data"
+)
+
+var (
+	ErrDataNotFound = errors.New("not found")
+)
+
+// MarshalData stores the source object as json data in the destination object annotations map.
+func MarshalData(src metav1.Object, dst metav1.Object) error {
+	data, err := json.Marshal(src)
+	if err != nil {
+		return err
+	}
+	if dst.GetAnnotations() == nil {
+		dst.SetAnnotations(map[string]string{})
+	}
+	dst.GetAnnotations()[DataAnnotation] = string(data)
+	return nil
+}
+
+// UnmarshalData tries to retrieve the data from the annotation and unmarshals it into the object passed as input.
+func UnmarshalData(from metav1.Object, to interface{}) error {
```

Comment:
I'm not crazy about the not-found error. Some possible alternatives:

- only returning an error if json unmarshaling failed (i.e. this function must not be called if the annotation is absent)
- returning `(bool, error)` with the `bool` indicating if the annotation was present (caller has to check `bool` status)

Or the least conventional, but one I kinda like because it doesn't require ifs/elses/not-found errors:

```go
func UnmarshalData(from metav1.Object, to interface{}, handler func()) error {
  data := from.GetAnnotations()[DataAnnotation]
  if data == "" {
    return nil
  }

  if err := json.Unmarshal([]byte(data), to); err != nil {
    return err
  }

  handler()

  return nil
}
```

And you'd use it like this:

```go
restored := &kubeadmbootstrapv1alpha3.KubeadmConfig{}
err := utilconversion.UnmarshalData(src, restored, func() {
  if restored.Status.DataSecretName != nil {
    dst.Status.DataSecretName = restored.Status.DataSecretName
  }
}
if err != nil {
 // ...
}
```

This relies on `restored` being in scope of the closure of the anonymous handler function. The handler function is only invoked if the marshaled data exists in the annotation. This may not be great go, and maybe there are problems with it I'm not thinking about, but I like not having to check errors or bools.

Reply:
Thanks for the suggestions!

I went with the second option, the bool seemed the least complicated and familiar to use if someone has used maps in go before.